mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala

```diff
@@ -0,0 +1,472 @@
+   *               data (token counts) in edge descriptors.
+   * @param k  Number of topics
+   * @param vocabSize  Number of unique terms
+   * @param topicSmoothing  "alpha"
+   * @param termSmoothing  "eta"
+   */
+  private[clustering] case class LearningState(
+      graph: Graph[TopicCounts, TokenCount],
+      k: Int,
+      vocabSize: Int,
+      topicSmoothing: Double,
+      termSmoothing: Double) {
+    // TODO: Checkpoint periodically?
+    def next(): LearningState = copy(graph = step(graph))
+
+    private def step(graph: Graph[TopicCounts, TokenCount]): Graph[TopicCounts, TokenCount] = {
+      val eta = termSmoothing
+      val W = vocabSize
+      val alpha = topicSmoothing
+
+      val N_k = collectTopicTotals()
+      val sendMsg: EdgeContext[TopicCounts, TokenCount, (Boolean, TopicCounts)] => Unit =
+        (edgeContext) => {
+          // Compute N_{wj} gamma_{wjk}
+          val N_wj = edgeContext.attr
+          // E-STEP: Compute gamma_{wjk} (smoothed topic distributions), scaled by token count
+          // N_{wj}.
+          val scaledTopicDistribution: TopicCounts =
+            computePTopic(edgeContext.srcAttr, edgeContext.dstAttr, N_k, W, eta, alpha) *= N_wj
+          edgeContext.sendToDst((false, scaledTopicDistribution))
+          edgeContext.sendToSrc((false, scaledTopicDistribution))
+        }
+      // This is a hack to detect whether we could modify the values in-place.
+      // TODO: Add zero/seqOp/combOp option to aggregateMessages.
+      val mergeMsg: ((Boolean, TopicCounts), (Boolean, TopicCounts)) => (Boolean, TopicCounts) =
+        (m0, m1) => {
+          val sum =
+            if (m0._1) {
+              m0._2 += m1._2
+            } else if (m1._1) {
+              m1._2 += m0._2
+            } else {
+              m0._2 + m1._2
+            }
+          (true, sum)
+        }
+      // M-STEP: Aggregation computes new N_{kj}, N_{wk} counts.
+      val docTopicDistributions: VertexRDD[TopicCounts] =
+        graph.aggregateMessages[(Boolean, TopicCounts)](sendMsg, mergeMsg)
+          .mapValues(_._2)
+      // Update the vertex descriptors with the new counts.
+      graph.outerJoinVertices(docTopicDistributions) { (vid, oldDist, newDist) => newDist.get }
+    }
+
+    def collectTopicTotals(): TopicCounts = {
+      val numTopics = k
+      graph.vertices.filter(isTermVertex).values.fold(BDV.zeros[Double](numTopics))(_ += _)
+    }
+
+    /**
+     * Compute the log likelihood of the observed tokens, given the current parameter estimates:
+     *  log P(docs | topics, topic distributions for docs, alpha, eta)
+     *
+     * Note:
+     *  - This excludes the prior; for that, use [[logPrior]].
+     *  - Even with [[logPrior]], this is NOT the same as the data log likelihood given the
+     *    hyperparameters.
+     */
+    lazy val logLikelihood: Double = {
+      val eta = termSmoothing
+      val alpha = topicSmoothing
+      assert(eta > 1.0)
```

Comment:
`termSmoothing` could be in `(0, 1)` given the current precondition checks. In the example code, the default is 0.1, which leads to assert error here.